### DISCUSSION TITLE
Route::resource() for Orchid

### DISCUSSION THREAD

**Discussion by kdv88** [orchidsoftware/platform#1613] [Ideas]
Laravel has the Route::resource(), which automatically builds the routes. For now the Orchid-way would be to create separate routes:

```
Route::screen('projects', ProjectListScreen::class)
    ->name('platform.project.list');
Route::screen('project/create', ProjectEditScreen::class)
    ->name('platform.project.create');
Route::screen('project/{project}/edit', ProjectEditScreen::class)
    ->name('platform.project.edit');
Route::screen('project/{project}', ProjectShowScreen::class)
    ->name('platform.project.show');
```

In Laravel these can be summarized by `Route::resource('projects', 'ProjectController');`

I would love to see a `Route::ScreenResource('project')` which automatically handles routing.
Perhaps automatically looking for `{$name}EditScreen::class` and `{$name}ListScreen::class`

**Comment by angularcircle** [comment]
If you'll use [Orchid CRUD](https://github.com/orchidsoftware/crud) package - it will handle default routes for you

**Reply by kdv88** [reply]
But how would I go if I want to utilise my own customised screens, but also want to have the benefits of a single line in the routes-file?

I like single `Route::resource('projects', 'ProjectController');` in my routes-file more, than multiple routes.
But I also want to use my own Crud-screens instead of the default.

**Comment by lemyskaman** [comment]
a route::screenResource() might not be a bad idea and I think is not so hard to implement 

Now it require to pas the listScreen and the Editscreen as param

Any way you can also point the CRUD route to the editScreen and, with some logic inside query method returns it along with the data. So wich layout to display (or the state of what ever you want to show should match your create or update or view or delete action to perform ).

**Reply by lemyskaman** [reply]
this is an exampple of what i do:

```php
class KlorchidTestScreen extends KlorchidCrudScreen
{

    /**
     * Display header name.
     *
     * @var string
     */
    public $name = 'KlorchidTestScreen';

    /**
     * Display header description.
     *
     * @var string|null
     */
    public $description = 'KlorchidTestScreen';




    public function modePermissions(): array
    {
        return [
            'view' => 'platform.klorchidtests.view',
            'edit' => 'platform.klorchidtests.edit',
            'create' => 'platform.klorchidtests.create',
        ];
    }

    /**
     * Query data.
     *
     * @return array
     */
    public function query(KlorchidTest $item): array
    {
        //$repository = $this->getRepository();
        //$mode = $this->setMode('create')->getMode();
        $mode = $this->setMode()->getMode();

         //\Debugbar::debug($repository->getModel()->klorchidTestType->name);
        return [

            model_keyname() => $item,
            'screen_mode' => $mode,
            'mode_perms' => $this->getModePerms(),
        ];
    }
```

**Reply by lemyskaman** [reply]
as you might guess $this->setMode()->getMode(); detect set and return the current Layout mode to display (based on my own logic)

And done. 

At the end remember that **a screen is a visual representation of some resource (I call resource to  "what the query method returns")** **so a screen will never be a resource** , for the contrary a controller conceptually talking can be a resource or a pre hall trhoug data might  be transformed or processed  from model before being delivered to response.

**Reply by kdv88** [reply]
Nice, that is taking it even further!

In my case I'm one hand looking for the Laravel routes:
```| admin           | POST                                   | projects                                            | admin.projects.store                         | App\Http\Controllers\Admin\ProjectController@store                          | admin         |
| admin           | GET|HEAD                               | projects                                            | admin.projects.index                         | App\Http\Controllers\Admin\ProjectController@index                          | admin         |
| admin           | GET|HEAD                               | projects/create                                     | admin.projects.create                        | App\Http\Controllers\Admin\ProjectController@create                         | admin         |
| admin           | GET|HEAD                               | projects/{project}                                  | admin.projects.show                          | App\Http\Controllers\Admin\ProjectController@show                           | admin         |
| admin           | DELETE                                 | projects/{project}                                  | admin.projects.destroy                       | App\Http\Controllers\Admin\ProjectController@destroy                        | admin         |
| admin           | PUT|PATCH                              | projects/{project}                                  | admin.projects.update                        | App\Http\Controllers\Admin\ProjectController@update                         | admin         |
| admin           | GET|HEAD                               | projects/{project}/edit                             | admin.projects.edit                          | App\Http\Controllers\Admin\ProjectController@edit                           | admin         |
```

And on the other hand I don't want to forget adding specific routes. So the Laravel `Route::resource()` does exactly what I want: give me a oneliner which creates simple routes.


> Now it require to pas the listScreen and the Editscreen as param

This is indeed the direction of what I am thinking. It is just a shortcut for multiple routes.

**Comment by tabuna** [comment]
Hi @kdv88. As mentioned earlier in the answer for a fast CRUD, we have a separate package. Maybe it will suit you for simple things. As for your proposal, I think it will be a little clear to users. Since `Route::resource()` expects only one class, and your proposal has 4 of them. As far as I know, this was even removed from the Laravel documentation (Correct me if it isn't)

It's effortless to do. You can do it yourself in your app or release it as a package. 
Just specify the macro for `Route` in the service provider:

```php
use Illuminate\Support\Facades\Route;

Route::macro('resourceScreen', function ($params) {
   // The logic you want to get
});
```

Then, you can call in your routes:

```php
Route::resourceScreen('Any values will be passed to the closure function')
```

It also makes it impossible to ask any additional things, like excluding middleware or breadcrumbs.


You can read more about macros here: https://laravel-news.com/laravel-route-tips-to-improve-your-routing or https://laravel.com/api/8.x/Illuminate/Support/Traits/Macroable.html (If you like reading code more)

**Reply by kdv88** [reply]
> Hi @kdv88. As mentioned earlier in the answer for a fast CRUD, we have a separate package. Maybe it will suit you for simple things. As for your proposal, I think it will be a little clear to users. Since `Route::resource()` expects only one class, and your proposal has 4 of them. As far as I know, this was even removed from the Laravel documentation (Correct me if it isn't)
> 

Still there:  https://laravel.com/docs/8.x/controllers#resource-controllers


> It's effortless to do. You can do it yourself in your app or release it as a package.
> Just specify the macro for `Route` in the service provider:
> 
> ```
> use Illuminate\Support\Facades\Route;
> 
> Route::macro('resourceScreen', function ($params) {
>    // The logic you want to get
> });
> ```
> 
> Then, you can call in your routes:
> 
> ```
> Route::resourceScreen('Any values will be passed to the closure function')
> ```

Perhaps this will be something to look at from my side.

Thank you!
